kernel/fcbfns.c

```diff
@@ -502,13 +502,39 @@ UBYTE FcbDelete(xfcb FAR * lpXfcb)
   return result;
 }
 
+static void FcbExpandAsterisks(char *p)
+{
+  int i, afound;
+
+  // Filename
+  for (i = 0, afound = 0; i < 8; i++) {
+    if (p[i] == '*' || afound) {
+      p[i] = '?';
+      afound = 1;
+    }
+  }
+
+  // Extension
+  for (i = 8, afound = 0; i < 11; i++) {
+    if (p[i] == '*' || afound) {
+      p[i] = '?';
+      afound = 1;
+    }
+  }
+}
+
 UBYTE FcbRename(xfcb FAR * lpXfcb)
 {
   rfcb FAR *lpRenameFcb;
   COUNT FcbDrive;
   UBYTE result = FCB_SUCCESS;
   void FAR *lpOldDta = dta;
 
+  /* DOS 3.0+ allows * wildcards for rename */
+  rfcb FAR *trfcb = (rfcb FAR *)lpXfcb;
```

Comment:
Why is such cast?
Are these structs compatible?

Reply:
Actually the pointer passed into FcbRename should be an `rfcb *`, but it's pushed through a generic parsing routine to get the first filename extracted as they are initially compatible. Later they cast it back to get to the second part of the structure. All pretty horrid if you ask me.

Reply:
There seem to be the `GetNameField()` in fcbfns.c,
which does similar to what you do?

Reply:
Seems to be operating with asciiz strings, we have 11 chars, no dot, and no terminator.

Reply:
It seems to be very generic: has `nFieldSize` for
non-asciiz. And more importantly, it is called from
`FcbParseFname()`, which, in turn, is called from
`FcbRename()`!

Reply:
If that's all working then perhaps it's just the destination pattern that's not being expanded?

Reply:
I bet.
Plus in this case you likely need to not only
expand, but also match with the from_name,
so that there no ? in the to_name at the end.